### PR TITLE
Add rules for atan2, hypot, mod, and rem. Addresses #15.

### PR DESCRIPTION
**Pull request by twadleigh**
Addresses #15.

### CODE REVIEW

File: src/rules.jl [hunk 1]
```diff
@@ -161,11 +161,10 @@ hasdiffrule(f::Symbol, arity::Int) = in((f, arity), DEFINED_DIFFRULES)
...
 
-# TODO:
-#
-# mod
-# hypot
-# atan2
+@define_diffrule atan2(x, y) = :($y / ($x*$x + $y*$y)), :(-$x / ($x*$x + $y*$y))
+@define_diffrule hypot(x, y) = :($x / hypot($x, $y)), :($y / hypot($x, $y))
+@define_diffrule mod(x, y)   = :(1), :(-trunc($x / $y))
+@define_diffrule rem(x, y)   = :(1), :(-trunc($x / $y))
```
Comment (twadleigh) [inline src/rules.jl]: These need fixing. `mod` and `rem` are not identical. Besides, contrary to my belief following an initial quick study of the cases, these rules don't work for all sign combinations for either function.

Comment (twadleigh) [inline src/rules.jl]: Nope. I was wrong about being wrong. :smile: The rule works for `rem`, but `trunc` needs to be changed to `floor` for `mod`.

**Comment by twadleigh** [comment]
The rules for `mod` and `rem` given here are valid for the partial derivatives when `x/y` is not an integer. They are valid on at least one side when `y != 0`.

Should they be modified to give `NaN` when `x/y` is integral? Are there cases where that sort of safeguard ends up being important (say, in optimizers)? My gut says that optimizers are going to have a hard time near discontinuities regardless, though.

**Comment by jrevels** [comment]
> Should they be modified to give NaN when x/y is integral

I think so, might as well have them be correct. We could change the rule to something like `:(z = $x / $y; first(promote(ifelse(isinteger(z), NaN, -floor(z)), NaN)))`, which avoids a branch and maintains type stability.

Thanks for this!

**Comment by twadleigh** [comment]
Oh, forgot to do the same for the first partial.

**Comment by twadleigh** [comment]
I'll squash shortly.

File: src/rules.jl [hunk 2]
```diff
@@ -161,11 +161,10 @@ hasdiffrule(f::Symbol, arity::Int) = in((f, arity), DEFINED_DIFFRULES)
 @define_diffrule /(x, y) = :(inv($y)),            :(-$x / ($y^2))
 @define_diffrule ^(x, y) = :($y * ($x^($y - 1))), :(($x^$y) * log($x))
 
-# TODO:
-#
-# mod
-# hypot
-# atan2
+@define_diffrule atan2(x, y) = :($y / ($x*$x + $y*$y)), :(-$x / ($x*$x + $y*$y))
```
Comment (jrevels) [inline src/rules.jl]: Can we switch `x * x` and `y * y` to `x^2` and `y^2`, just to keep in style with the other rules? Thanks!

Comment (twadleigh) [inline src/rules.jl]: K

**Pull request merged**